File: ocr.py
import sys
import Quartz
import Vision
from Cocoa import NSURL

def extract_text(image_path):
    input_url = NSURL.fileURLWithPath_(image_path)
    
    request_handler = Vision.VNImageRequestHandler.alloc().initWithURL_options_(input_url, None)
    
    results = []
    def completion_handler(request, error):
        if error:
            print(f"Error: {error}")
            return
        observations = request.results()
        for observation in observations:
            top_candidate = observation.topCandidates_(1).firstObject()
            if top_candidate:
                results.append(top_candidate.string())
                
    request = Vision.VNRecognizeTextRequest.alloc().initWithCompletionHandler_(completion_handler)
    request.setRecognitionLevel_(Vision.VNRequestTextRecognitionLevelAccurate)
    
    success, error = request_handler.performRequests_error_([request], None)
    if not success:
        print(f"Failed to perform request: {error}")
    
    print("\n".join(results))

extract_text(sys.argv[1])

File: ocr.swift
import Cocoa
import Vision

let args = CommandLine.arguments
if args.count < 2 {
    print("Usage: swift ocr.swift <image_path>")
    exit(1)
}

let imagePath = args[1]
guard let image = NSImage(contentsOfFile: imagePath),
      let tiffData = image.tiffRepresentation,
      let bitmap = NSBitmapImageRep(data: tiffData),
      let cgImage = bitmap.cgImage else {
    print("Failed to load image.")
    exit(1)
}

let requestHandler = VNImageRequestHandler(cgImage: cgImage, options: .init())
let request = VNRecognizeTextRequest { request, error in
    guard let observations = request.results as? [VNRecognizedTextObservation] else {
        print("No results")
        return
    }
    let recognizedStrings = observations.compactMap { observation in
        return observation.topCandidates(1).first?.string
    }
    print("--- OCR RESULTS ---")
    print(recognizedStrings.joined(separator: "\n"))
    print("-------------------")
}
request.recognitionLevel = .accurate

do {
    try requestHandler.perform([request])
} catch {
    print("Error: \(error)")
}
